### ISSUE TITLE
[Spec] Gate Modality Differentiation via Auxiliary Routing Loss

### ISSUE THREAD

**Issue by wray-lee** [wray-lee/NSMoR#12]
# Gate Modality Differentiation Spec

## Problem Statement

The MoR Router's `g_lif` gate no longer collapses (mean 0.428 ± 0.071 after lambda_reg fix), but it has not learned **condition-specific routing**. Trial-wise standard deviation remains small (0.071), meaning the gate settles into a compromise position rather than specializing by stimulus modality. The intended behavior — "wind trials route through LIF fast pathway, visual trials route through GRU smooth pathway" — is absent.

This prevents the model from exploiting the architectural separation hypothesis: LIF should handle rapid transients (wind onset, whisker contact) while GRU integrates slow looming visual expansion.

## Solution

Introduce an **auxiliary routing loss** that encourages trial-level gate differentiation by stimulus condition, while preserving the anti-collapse regularizer. The loss penalizes gate distributions that fail to separate across conditions, pushing wind-dominant trials toward high `g_lif` and visual-dominant trials toward high `g_gru`.

A secondary hypothesis: GRU pathway capacity may be insufficient to represent smooth visual trajectories, causing the router to fall back to a 50/50 blend. This spec focuses on the loss-driven differentiation first; capacity expansion (wider GRU hidden dim, deeper GRU stack) is out of scope but noted for future work.

## User Stories

1. As a computational neuroscience researcher, I want the MoR Router to assign high `g_lif` weight to wind-onset trials, so that the model's routing behavior reflects the biological hypothesis that fast mechanosensory transients engage the giant-fiber escape pathway.

2. As a model developer, I want a scalar diagnostic metric (per-condition gate mean, between-condition separation) logged during training, so that I can track whether the router is learning modality-specific policies without manually inspecting gate distributions.

3. As a model developer, I want the auxiliary routing loss to be optional (controlled by a `lambda_routing_aux` config parameter), so that I can ablate its effect by setting it to 0.0 and comparing against the baseline anti-collapse-only regime.

4. As a training script user, I want the auxiliary loss to respect the same warmup schedule as other bio-loss terms (energy/sparse/jerk), so that the router is not forced into premature specialization before the MSE loss has stabilized.

5. As a model developer, I want the auxiliary loss to operate on trial-level aggregated gates (mean `g_lif` over valid frames per trial), so that it does not penalize legitimate frame-by-frame routing variation within a single trial.

6. As a model developer, I want the loss to distinguish between **pure-wind trials** (no visual looming, 570 frames zero-padded) and **visual+wind trials** (looming present), so that the routing signal is aligned with the stimulus modality, not the behavioral outcome label.

7. As a model developer, I want the loss to be robust to trials where the stimulus modality is ambiguous (e.g., visual-only trials with no wind, or trials where looming and wind onsets coincide), so that the penalty does not fire on edge cases where no clear routing preference exists.

8. As a validation script user, I want `analyze_gating.py` to report **per-condition gate statistics** (mean `g_lif` for pure-wind vs visual-dominant trials), so that I can verify differentiation has occurred without re-implementing the diagnostic logic.

9. As a researcher, I want the auxiliary loss to be mathematically independent of the anti-collapse `lambda_reg` term, so that the two objectives (prevent collapse, encourage differentiation) can be tuned separately.

10. As a model developer, I want the auxiliary loss gradient to flow through the MoR Router only (not the LIF or GRU pathways), so that pathway internals remain uncorrupted by the routing signal.

11. As a training script user, I want a clear log message at epoch start showing `lambda_routing_aux` (both raw config value and warmup-scaled effective value), so that I can audit whether the term is active and at what magnitude.

12. As a model developer, I want the loss to degrade gracefully if the dataset contains no pure-wind trials, so that the training script does not crash or produce NaN when applied to a visual-only corpus.

### Loss Function Design

The auxiliary routing loss operates on **trial-level aggregated gates**, not frame-level gates. For each trial, compute the mean `g_lif` over valid (non-padded) frames. Group trials by **stimulus modality** (pure-wind vs visual-present), then penalize the **overlap** between the two distributions.

**Candidate formulation** (to be prototyped):
```python
# Per-trial gate aggregation
g_lif_per_trial = masked_mean(g_lif, lengths)  # (B,)

# Partition by modality (requires a `wind_only_mask` boolean tensor per trial)
g_wind = g_lif_per_trial[wind_only_mask]
g_visual = g_lif_per_trial[~wind_only_mask]

# Separation objective: maximize (mean(g_wind) - mean(g_visual))
# or minimize overlap via hinge loss
loss_routing_aux = max(0, margin - (g_wind.mean() - g_visual.mean()))
```

The `margin` hyperparameter (e.g., 0.2) defines the minimum desired separation. If `mean(g_wind) - mean(g_visual) < margin`, the loss fires; otherwise it is zero.

Alternative: use a **contrastive** or **triplet-style** loss that pulls same-modality trials together and pushes cross-modality trials apart in gate space. This is more complex but may generalize better to multi-modal mixtures.

**Decision**: Start with the simpler hinge-based separation loss. If it fails to induce differentiation, escalate to contrastive.

### Stimulus Modality Signal

The model must distinguish **pure-wind trials** from **visual-present trials**. This information is not currently exposed in the dataset schema as a per-trial boolean flag.

**Option A**: Derive from `Snapshot.wind_state` — pure-wind trials have `wind_state=1` and zero visual looming (inferred from `visual_angle=0` or `looming_velocity=0` at anchor).

**Option B**: Add an explicit `is_pure_wind` boolean flag to the dataset schema during ETL (`prepare_data.py`), stamped at the same time as the MCMC Prior.

**Decision**: Option B. The flag is cheap to compute during ETL (one-time cost), avoids runtime heuristics, and is forwards-compatible if the corpus later includes multi-modal trials where `wind_state=1` does not imply absence of looming.

### Integration into `BioJointLoss`

The auxiliary loss is a **fourth bio-loss term**, similar in structure to `lambda_energy`, `lambda_sparse`, and `lambda_jerk`. It receives the warmup `annealing_factor` like the others, so it ramps from 0 over the first `warmup_epochs`.

**Signature change** (no breaking change — new parameter is optional with default 0.0):
```python
def forward(
    self,
    y_pred: torch.Tensor,
    y_true: torch.Tensor,
    lengths: torch.Tensor,
    g_gru: torch.Tensor,
    lif_spikes: Optional[torch.Tensor] = None,
    *,
    lambda_reg: float = 0.2,
    lambda_energy: float = 0.0,
    lambda_sparse: float = 0.0,
    lambda_jerk: float = 0.0,
    lambda_routing_aux: float = 0.0,  # NEW
    wind_only_mask: Optional[torch.Tensor] = None,  # NEW, shape (B,)
    annealing_factor: float = 1.0,
) -> torch.Tensor:
    ...
    if lambda_routing_aux > 0.0 and wind_only_mask is not None:
        routing_aux_loss = compute_routing_aux_loss(
            g_lif, lengths, wind_only_mask, margin=0.2
        )
        loss = loss + lambda_routing_aux * annealing_factor * routing_aux_loss
    return loss
```

The `wind_only_mask` tensor is passed from the training loop, which extracts it from a new dataset key `is_pure_wind` (boolean array, shape `(n_trials,)`).

### Dataset Schema Extension

Add `is_pure_wind: np.ndarray[bool]` to `nsmor_dataset.pt` schema, generated in `prepare_data.py` by checking `Snapshot.wind_state == 1` and `Snapshot.visual_angle == 0.0`.

**Migration**: Existing datasets without this key are compatible — `BioJointLoss` will skip the auxiliary term if `wind_only_mask is None`. No version bump required; this is an additive change.

### Configuration

Add `lambda_routing_aux: float = 0.0` to `LossConfig` in `nsmor/config_parser.py` and `config/default.yaml`. Default is 0.0 (disabled), so the baseline anti-collapse behavior is unchanged.

### Validation Diagnostic

Extend `analyze_gating.py` to report:
- Mean `g_lif` for pure-wind trials
- Mean `g_lif` for visual-present trials
- Absolute separation: `|mean(g_wind) - mean(g_visual)|`

This metric is computed on the validation set and printed alongside existing gate stats.

### Gradient Flow Isolation

The auxiliary loss computes gradients only with respect to the MoR Router weights. LIF and GRU pathway weights must not be updated by this term. This is achieved by:
1. The loss operates on `g_lif` (router output), not on LIF/GRU hidden states.
2. The router is a separate submodule (`backend.routing_net`), so PyTorch's autograd naturally isolates the gradient.

No explicit `detach()` is required if the loss is defined correctly.

### Out of Scope

- **GRU capacity expansion**: Increasing `hidden_dim` or adding a second GRU layer to test whether the lack of differentiation is due to representational bottleneck, not routing signal. This is a separate experiment.
- **Multi-condition routing**: Extending the loss to handle more than two modality classes (e.g., wind-only, visual-only, wind+visual, whisker-only). The current corpus has only wind and visual stimuli.
- **Frame-level routing penalties**: Penalizing rapid gate switches within a trial (smoothness regularizer). This is orthogonal to trial-level differentiation.
- **Routing visualization**: Generating per-trial gate heatmaps or t-SNE projections of gate distributions. This is analysis tooling, not core model behavior.

### Test Seams

**Primary seam**: `BioJointLoss.forward()` — unit test the auxiliary loss computation in isolation, verifying that:
1. When `wind_only_mask` partitions trials into two groups with distinct gate means, the loss is zero if separation exceeds `margin`, and positive otherwise.
2. When `wind_only_mask` is `None`, the loss is zero regardless of `lambda_routing_aux`.
3. When `lambda_routing_aux=0.0`, the loss contribution is zero.

**Secondary seam**: `train_one_epoch()` integration — end-to-end smoke test that:
1. A 2-epoch training run with `lambda_routing_aux > 0` and a synthetic dataset containing `is_pure_wind` completes without NaN/Inf.
2. The training log records the `lambda_routing_aux` value.
3. The loss decreases (sanity check — not a proof of differentiation, just that the term is active).

**Tertiary seam**: `analyze_gating.py` diagnostic — verify that the per-condition gate statistics are computed and printed correctly.

### Prior Art

- `tests/test_loss.py::test_bio_joint_loss_validates_shapes` — similar structure for testing loss shape contracts.
- `tests/test_lambda_reg_warmup.py::test_lambda_reg_not_warmup_scaled_in_train_loop` — mocking `train_one_epoch` to observe parameter propagation.
- `tests/test_pipeline.py::test_prepare_data_smoke` — ETL smoke test pattern for schema extensions.

### Good Test Properties

- **External behavior only**: Test the loss output value and gradient flow, not internal intermediate tensors.
- **Deterministic**: Fix random seeds for any sampling in the test dataset.
- **Minimal synthetic data**: Use 10-20 trials, not the full 288-trial subset, to keep test runtime under 5 seconds.

## Further Notes

- The `margin=0.2` hyperparameter is a starting guess. If differentiation is weak, it may need to be tuned (via grid search or manual iteration) or replaced with a learned temperature parameter.
- If the hinge loss fails to induce separation, the next escalation is a **contrastive loss** (InfoNCE-style) that treats same-modality trials as positive pairs and cross-modality trials as negatives. This requires more implementation complexity (sampling pairs, temperature tuning) but has stronger theoretical grounding.
- The user noted that wind channels in the sensory encoder have only 0.4% non-zero activation. This may be a separate bottleneck: if the wind signal never reaches the MoR Router, no amount of routing loss will fix it. A diagnostic pass (measure wind channel activation distribution) should be added to `analyze_gating.py` to rule this out.

**Comment by wray-lee** [comment]
Spec complete: all tickets (#13–#18) delivered and verified. Ticket #13 (ETL is_pure_wind) was already present. Tickets #14 (loss), #15 (config), #16 (training loop), #17 (analysis), #18 (E2E test) all implemented and CI passed (run #83). Auxiliary routing loss for gate modality differentiation is now fully integrated.
